src/ir/resources.rs

```diff
@@ -1 +1,297 @@
+use crate::ir::reference::{Origin, Reference};
+use crate::parser::resource::ResourceValue;
+use crate::parser::sub::{sub_parse_tree, SubValue};
+use crate::specification::{spec, Complexity, Specification};
+use crate::{CloudformationParseTree, TransmuteError};
+use std::collections::HashMap;
 
+// ResourceIr is the intermediate representation of a nested stack resource.
+// It is slightly more refined than the ResourceValue, in some cases always resolving
+// known types. It also decorates objects with the necessary information for a separate
+// system to output all the necessary internal structures appropriately.
+#[derive(Clone)]
+pub enum ResourceIr {
+    Null,
+    Bool(bool),
+    Number(i64),
+    String(String),
+
+    // Higher level resolutions
+    Array(Complexity, Vec<ResourceIr>),
+    Object(Complexity, HashMap<String, ResourceIr>),
+
+    /// Rest is meta functions
+    /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html#w2ab1c33c28c21c29
+    If(String, Box<ResourceIr>, Box<ResourceIr>),
+    Join(String, Vec<ResourceIr>),
+    Ref(Reference),
+    GetAtt(String, String),
+    Sub(Vec<ResourceIr>),
+    Map(Box<ResourceIr>, Box<ResourceIr>, Box<ResourceIr>),
+}
+
+/// ResourceTranslationInputs is a place to store all the intermediate recursion
+/// for resource types.
+#[derive(Clone, Debug)]
+pub struct ResourceTranslationInputs<'t> {
+    parse_tree: &'t CloudformationParseTree,
+    specification: &'t Specification,
+    complexity: Complexity,
+    property_type: Option<&'t str>,
+    resource_type: &'t str,
+}
+
+// ResourceInstruction is all the information needed to output a resource assignment.
+pub struct ResourceInstruction {
+    pub name: String,
+    pub condition: Option<String>,
+    pub resource_type: String,
+    pub properties: HashMap<String, ResourceIr>,
+}
+
+pub fn translates_resources(parse_tree: &CloudformationParseTree) -> Vec<ResourceInstruction> {
+    let spec = spec();
+    let mut resource_instructions = Vec::new();
+    for resource in parse_tree.resources.resources.iter() {
+        let resource_spec = spec
```

Comment:
Nit: This is more of an optimization question. Does this need to be done for every iteration? Can we save resource_spec after the first time we see the type?

Reply:
I'm a bit confused, as each resource is unique, and the lookup is O(1) already (as it's backed by a hashmap in spec).